### ISSUE TITLE
Warnings when importing on macOS

### ISSUE THREAD

**Issue by saraedum** [conda-forge/cppyy-feedstock#12]
On macOS I get the following warning when importing cppyy:
```
    Error in <UnknownClass::InitInterpreter()>: LLVM SYMBOLS ARE EXPOSED TO CLING! This will cause problems; please hide them or dlopen() them after the call to TROOT::InitInterpreter()!
```

**Comment by saraedum** [comment]
@wlav, is this something to worry about?

**Comment by wlav** [comment]
No; I actually saw that earlier and asked upstream. First, there is this discussion:
 https://github.com/conda-forge/root-feedstock/pull/26

with this change on their side:
 https://github.com/conda-forge/root-feedstock/blob/f6148459d19e0122ebfd5fb66fc27d1479a8088f/recipe/build_root-base.sh#L30-L34

Second, Vassil told me to make sure to load libCling with RTLD_LOCAL, which is a simple patch in loader.py:

```diff --git a/cling/python/cppyy_backend/loader.py b/cling/python/cppyy_backend/loader.py
index 3943f44..2c2ba91 100644
--- a/cling/python/cppyy_backend/loader.py
+++ b/cling/python/cppyy_backend/loader.py
@@ -36,8 +36,9 @@ def _load_helper(bkname):
             for loc in ['lib', 'bin']:
                 fpath = os.path.join(pkgpath, loc, dep+soext)
                 if os.path.exists(fpath):
-                    dep = fpath
-                    ctypes.CDLL(dep, ctypes.RTLD_GLOBAL)
+                    ldtype = ctypes.RTLD_GLOBAL
+                    if dep == 'libCling': ldtype = ctypes.RTLD_LOCAL
+                    ctypes.CDLL(fpath, ldtype)
                     break
         return ctypes.CDLL(os.path.join(pkgpath, 'lib', bkname), ctypes.RTLD_GLOBAL)
     except OSError:
```
and actually solves the issue. (I still need to test it on the other platforms.)

Finally, when testing, I didn't see any bad side effects either which way.

There are other issues with using stock cling, mostly related to performance (as much as 50000x in one case, so worth it), but that's on me to get those either upstreamed, or solve them differently.

**Comment by wlav** [comment]
Actually no wait, checking the build file and it's stock clang, stock llvm, but cling from cppyy-cling, so it does have the performance patches.

I have one patch on clang for performance, but that's Linux-only anyway. (Mac would benefit also, but it's an Apple thing, so safe to remove on Linux, not on Mac.)

**Comment by saraedum** [comment]
> Second, Vassil told me to make sure to load libCling with RTLD_LOCAL, which is a simple patch in loader.py: 
> […]
> and actually solves the issue. (I still need to test it on the other platforms.)

Ok. So I'll just wait for cppyy-cling to have that patch.

> There are other issues with using stock cling, mostly related to performance (as much as 50000x in one case, so worth it), but that's on me to get those either upstreamed, or solve them differently.

Hm…not sure I understand. We are using cppyy-cling here not stock cling. (Ah, as you noted below :)

**Comment by saraedum** [comment]
> I have one patch on clang for performance, but that's Linux-only anyway. (Mac would benefit also, but it's an Apple thing, so safe to remove on Linux, not on Mac.)

Which one is that? So I can propose it for our clang?

**Comment by wlav** [comment]
My memory is failing, I actually have 3 fixes for Clang, not just one:

For printing templated 'operator<' methods:
  https://bitbucket.org/wlav/cppyy-backend/src/master/cling/patches/clang_printing.diff

Part of the following, to prevent instantiation of templates that are declared 'extern'. On Linux, this gets thrown away by the linker, on Mac it's part of a security hardening feature (prevents you from pre-loading standard library symbols). For Cling, it's costly when templates are complicated:
  https://bitbucket.org/wlav/cppyy-backend/src/master/cling/patches/explicit_template.diff

And deep inside this one, there is a fix for name mangling of global variables on Windows CodeGen. This is actually fixed in later Clangs:
  https://bitbucket.org/wlav/cppyy-backend/src/master/cling/patches/win64s2.diff

Aside, the promise from upstream is to move to a newer clang after 6.20 is out. I really hope so, b/c Clang5 is getting long in the tooth and failing on some legal C++17 templates that gcc7 munches w/o problem. There has also been good progress in Windows support; hopefully a newer clang will resolve some remaining problems that I have, such as with ostreams.
